test/compress/pure_funcs.js

```diff
@@ -293,3 +293,16 @@ unary: {
         bar();
     }
 }
+
+issue_2629: {
+    options = {
+        side_effects: true,
+    }
+    input: {
+        /*@__PURE__*/ f1();
+        (/*@__PURE__*/ f2)();
```

Comment:
I don't think we want to drop this. The pure annotation only presently works for calls. Here the symbol `f2` is annotated as pure, and it is not dropped in v3.2.2 when a symbol:
```js
$ echo 'a; /*@__PURE__*/b; c;' | bin/uglifyjs -c
a,b,c;

$ echo 'a(); /*@__PURE__*/b(); c();' | bin/uglifyjs -c
a(),c();
```

Reply:
There is no distinction between annotating on `f2` or `f2()` in this case – they are one and the same.

See https://github.com/mishoo/UglifyJS2/pull/2633/files#diff-2cef7371337131fb81c6b3880549b93eR211

Reply:
I appreciate there's no distinction in this PR, but that wasn't the case in v3.2.2.

We could extend the notion of pure annotations to cover function expressions and function symbols, but originally it was only intended for calls specifically.

But we have to be logically consistent with respect to inlining values and functions.

With this PR:
```js
$ cat testp.js 
!function(){
    /*@__PURE__*/function bar() {
        console.log("bar");
    }
    bar();

    ( /*@__PURE__*/ function foo() { console.log("foo"); } )();
}();
```
```js
$ bin/uglifyjs testp.js -c
!function(){console.log("bar")}();
```
Why is `foo()` dropped but `bar()` kept?

Reply:
I don't see how this will ever be confusing - `(/*@__PURE__*/f2)` still won't be dropped after this PR. You need it to be `AST_Call.expression` in order for the pure annotation to be effective.

Reply:
It's not logically consistent. See test above. Why is one dropped and one retained?

If you can rig it such that *functions* can be annotated as pure in the `function bar` case above and then every single future `bar()` call will be dropped then it would logically consistent.

Reply:
> Why is one dropped and one retained?

Because one is an IIFE while the other one isn't - I don't see any confusion there whatsoever.

Reply:
But once `function bar` is inlined in `bar()` why won't it carry the pure annotation comment with it and behave exactly like the IIFE where the `AST_Call.expression` is annotated?  Once inlining is taken into account it becomes logically inconsistent.